graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlCache.java

```diff
@@ -232,70 +220,99 @@ public void onKeyspaceChanged(
   }
 
   /**
-   * Holds the GraphQL schema for a particular keyspace (either CQL-first or GraphQL-first,
-   * depending on whether there is a SchemaSource for this keyspace).
+   * An entry that holds the GraphQL schema cached for a particular keyspace (either CQL-first or
+   * GraphQL-first, depending on whether a custom schema was deployed).
    */
-  class DmlGraphqlHolder {
+  interface GraphqlHolder {
+    GraphQL getGraphql();
 
-    // the user-provided GraphQL schema, or null if CQL-first
-    private final SchemaSource source;
-    private final Keyspace keyspace;
-    // This future is used to guarantee that the cached value is built only once, even if two
-    // clients try to initialize it concurrently.
-    private final CompletableFuture<GraphQL> graphqlFuture = new CompletableFuture<>();
+    /**
+     * Whether the schema matches the given source (if it doesn't, it will have to be recomputed).
+     * Note that the source can be {@code null} to indicate that no custom schema was deployed.
+     */
+    boolean matches(@Nullable SchemaSource source);
 
-    DmlGraphqlHolder(SchemaSource source, Keyspace keyspace) {
-      this.source = source;
-      this.keyspace = keyspace;
-    }
+    boolean isCqlFirst();
+  }
 
-    DmlGraphqlHolder(SchemaSource source, GraphQL graphQL) {
+  /**
+   * An entry used when we need to recompute the schema (because we're reacting to the insertion of
+   * a new row in the source table).
+   */
+  class LazyGraphqlHolder implements GraphqlHolder {
+    private final SchemaSource source;
+    private final Supplier<GraphQL> computeGraphql;
```

Comment:
Why we don't use `CompletableFuture`? 
I think that it may be cleaner, and also we may not need explicit synchronization:
https://github.com/stargate/stargate/pull/971/files#diff-9e565971d8098ad42c987560f96a7f22a8ca0ae1987d11e9c2ebfee6d87ee84eR268

Reply:
Lazy init is a bit simpler than the two-step "put / init" mechanism we had before. We Ivan's suggestion of using Guava memoizer, we don't need the explicit synchronization anymore.